### PR TITLE
Supporting replication of TTL updates

### PR DESCRIPTION
**Pull request by vgkholla**
Will replicate the TTL update record as long as
1) There is no delete (either in remote or local)
2) The ID is not deprecated

If PUT is present locally, applies the TTL update
If PUT is not present, replicates the PUT and then immediately applies the TTL update

### CODE REVIEW

**Comment by vgkholla** [comment]
@dharju reminder to review

**Comment by vgkholla** [comment]
@pnarayanan @dharju reminder to review

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java [hunk 1]
```diff
@@ -763,6 +768,9 @@ private void writeMessagesToLocalStoreAndAdvanceTokens(List<ExchangeMetadataResp
                   notification.onBlobReplicaCreated(dataNodeId.getHostname(), dataNodeId.getPort(),
                       messageInfo.getStoreKey().getID(), BlobReplicaSourceType.REPAIRED);
                 }
+                if (messageInfo.isTtlUpdated()) {
```
Comment (zzmao) [inline ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: Why not handle this case  in `processReplicaMetadataResponse`? 
The `else` block from line 587 to 609 looks like the right place.

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java [hunk 2]
```diff
@@ -547,12 +543,13 @@ private void processReplicaMetadataResponse(Set<StoreKey> missingRemoteStoreKeys
       } else if (!missingRemoteStoreKeys.contains(messageInfo.getStoreKey())) {
         // the key is present in the local store. Mark it for deletion if it is deleted in the remote store and not
         // deleted yet locally
-        if (messageInfo.isDeleted() && !remoteReplicaInfo.getLocalStore().isKeyDeleted(localKey)) {
+        boolean deletedLocally = remoteReplicaInfo.getLocalStore().isKeyDeleted(localKey);
+        if (messageInfo.isDeleted() && !deletedLocally) {
           MessageFormatInputStream deleteStream =
```
Comment (zzmao) [inline ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: New a method `applyDelete()` ?

Comment (vgkholla) [reply ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: This class probably needs a refactor (for e.g., it uses blocking I/O). I don't want to modify anything that already exists in the context of this change.

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java (hunk 1 above)
Comment (vgkholla) [reply ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: The blob is not yet present at that point. This TTL update is applied after the blob is put into the store.

Comment (zzmao) [reply ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: I see. It 'may' not yet present, right?

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java [hunk 3]
```diff
@@ -799,6 +807,52 @@ private void writeMessagesToLocalStoreAndAdvanceTokens(List<ExchangeMetadataResp
...
+        new MessageInfo(messageInfo.getStoreKey(), ttlUpdateStream.getSize(), false, true, messageInfo.getAccountId(),
+            messageInfo.getContainerId(), messageInfo.getOperationTimeMs());
+    MessageFormatWriteSet writeSet = new MessageFormatWriteSet(ttlUpdateStream, Collections.singletonList(info), false);
+    DataNodeId remoteNode = remoteReplicaInfo.getReplicaId().getDataNodeId();
+    try {
+      // NOTE: It is possible that the key in question may have expired and this TTL update is being applied after it
+      // is deemed expired. The store will accept the op (BlobStore looks at whether the op was valid to do at the time
+      // of the op, not current time) but if compaction is running at the same time and has decided to clean up the
+      // record before this ttl update was applied (and this didn't find the key missing because compaction has not yet
+      // committed), then we have a bad situation where only a TTL update exists in the store. This problem has to be
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: Yup - it will be removed because there is no "anchoring" put.

File: ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java [hunk 4]
```diff
@@ -488,6 +489,230 @@ public void replicaThreadTestConverter() throws Exception {
...
+      addTtlUpdateMessagesToReplicasOfPartition(pid, b1, remoteHostOnly);
+      addTtlUpdateMessagesToReplicasOfPartition(pid, b1p, expectedLocalHostOnly);
+      addTtlUpdateMessagesToReplicasOfPartition(pid, b2, remoteHostOnly);
+      addTtlUpdateMessagesToReplicasOfPartition(pid, b2p, expectedLocalHostOnly);
+      addTtlUpdateMessagesToReplicasOfPartition(pid, b3, remoteHostOnly);
+
+      numMessagesInEachPart = remoteHost.infosByPartition.get(pid).size();
+    }
+
+    Properties properties = new Properties();
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java]: That's a good idea. I'll add that as a last commit before this code is merged (or we can add it after the merge). I don't want to make the change now because it will make the diff larger

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java [hunk 5]
```diff
@@ -580,6 +577,12 @@ private void processReplicaMetadataResponse(Set<StoreKey> missingRemoteStoreKeys
             notification.onBlobReplicaDeleted(dataNodeId.getHostname(), dataNodeId.getPort(), localKey.getID(),
                 BlobReplicaSourceType.REPAIRED);
           }
+        } else if (!deletedLocally && !messageInfo.isDeleted() && messageInfo.isTtlUpdated()) {
```
Comment (pnarayanan) [inline ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: Can `messageInfo.isDeleted()` and `messageInfo.isTtlUpdated()` both be true? Couldn't this check be `... else if (messageInfo.isTtlUpdated() && !deletedLocally) {}`?

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java (hunk 1 above)
Comment (pnarayanan) [reply ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: Any value in doing it this way, rather than constructing a single "compacted" record and putting it?

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pnarayanan) [reply ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: A stale TTL update means that the blob is permanent in other stores though, right? What is the stand on that? That it will only happen in rare scenarios?

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java (hunk 5 above)
Comment (vgkholla) [reply ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: Yes it can be both ttl updated and deleted (just like an `IndexValue`) - the flags are additive. I think it is useful to store the history this way.

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java (hunk 1 above)
Comment (vgkholla) [reply ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: I thought about that but that would basically be have to be done by a `Transformer` like entity (for e.g. I could code this into the `BlobIdTransformer`) but I those impls are pluggable and removing that transformer or substituting will affect correctness in a way that is not obvious. So I decided to it this way since this is independent of any `Transformer` impls. If there is a reliable way of compacting this (without creating more copies in memory), I can consider that (in this or a subsequent PR)

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: Yes - will happen only on rare scenarios and is a corner case that I need to give some thought to (probably find a way to delay compaction if the store is "too behind).

File: ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java [hunk 6]
```diff
@@ -488,6 +489,230 @@ public void replicaThreadTestConverter() throws Exception {
...
+    StoreKeyFactory storeKeyFactory = new BlobIdFactory(clusterMap);
+    Transformer transformer = new BlobIdTransformer(storeKeyFactory, storeKeyConverter);
+
+    List<PartitionId> partitionIds = clusterMap.getWritablePartitionIds(null);
+    int numMessagesInEachPart = 0;
+    Map<PartitionId, StoreKey> idsDeletedLocallyByPartition = new HashMap<>();
+    List<Host> remoteHostOnly = Collections.singletonList(remoteHost);
+    List<Host> expectedLocalHostOnly = Collections.singletonList(expectedLocalHost);
+    List<Host> localHostAndExpectedLocalHost = Arrays.asList(localHost, expectedLocalHost);
+    List<Host> remoteHostAndExpectedLocalHost = Arrays.asList(expectedLocalHost, remoteHost);
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java]: Done

**Comment by vgkholla** [comment]
@dharju @pnarayanan Comments addressed

**Comment by vgkholla** [comment]
@pnarayanan @dharju Reminder to review

File: ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java]: Addressed this
